File: lua/test/associated_person_entity_test.lua
-- AssociatedPerson entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("kota_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("AssociatedPersonEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:AssociatedPerson(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["associated_person"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:AssociatedPerson(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:AssociatedPerson(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = associated_person_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "associated_person." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set KOTA_TEST_ASSOCIATED_PERSON_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local associated_person_ref01_ent = client:AssociatedPerson(nil)
    local associated_person_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.associated_person"), "associated_person_ref01"))
    associated_person_ref01_data["employee_id"] = setup.idmap["employee01"]

    local associated_person_ref01_data_result, err = associated_person_ref01_ent:create(associated_person_ref01_data, nil)
    assert.is_nil(err)
    associated_person_ref01_data = helpers.to_map(associated_person_ref01_data_result)
    assert.is_not_nil(associated_person_ref01_data)
    assert.is_not_nil(associated_person_ref01_data["id"])

    -- LIST
    local associated_person_ref01_match = {
      ["employee_id"] = setup.idmap["employee01"],
    }

    local associated_person_ref01_list_result, err = associated_person_ref01_ent:list(associated_person_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(associated_person_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(associated_person_ref01_list_result),
      { id = associated_person_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local associated_person_ref01_data_up0_up = {
      id = associated_person_ref01_data["id"],
      ["employee_id"] = setup.idmap["employee_id"],
    }

    local associated_person_ref01_markdef_up0_name = "date_of_birth"
    local associated_person_ref01_markdef_up0_value = "Mark01-associated_person_ref01_" .. tostring(setup.now)
    associated_person_ref01_data_up0_up[associated_person_ref01_markdef_up0_name] = associated_person_ref01_markdef_up0_value

    local associated_person_ref01_resdata_up0_result, err = associated_person_ref01_ent:update(associated_person_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local associated_person_ref01_resdata_up0 = helpers.to_map(associated_person_ref01_resdata_up0_result)
    assert.is_not_nil(associated_person_ref01_resdata_up0)
    assert.are.equal(associated_person_ref01_resdata_up0["id"], associated_person_ref01_data_up0_up["id"])
    assert.are.equal(associated_person_ref01_resdata_up0[associated_person_ref01_markdef_up0_name], associated_person_ref01_markdef_up0_value)

    -- LOAD
    local associated_person_ref01_match_dt0 = {
      id = associated_person_ref01_data["id"],
    }
    local associated_person_ref01_data_dt0_loaded, err = associated_person_ref01_ent:load(associated_person_ref01_match_dt0, nil)
    assert.is_nil(err)
    local associated_person_ref01_data_dt0_load_result = helpers.to_map(associated_person_ref01_data_dt0_loaded)
    assert.is_not_nil(associated_person_ref01_data_dt0_load_result)
    assert.are.equal(associated_person_ref01_data_dt0_load_result["id"], associated_person_ref01_data["id"])

    -- REMOVE
    local associated_person_ref01_match_rm0 = {
      id = associated_person_ref01_data["id"],
    }
    local _, err = associated_person_ref01_ent:remove(associated_person_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local associated_person_ref01_match_rt0 = {
      ["employee_id"] = setup.idmap["employee01"],
    }

    local associated_person_ref01_list_rt0_result, err = associated_person_ref01_ent:list(associated_person_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(associated_person_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(associated_person_ref01_list_rt0_result),
      { id = associated_person_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function associated_person_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/associated_person/AssociatedPersonTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read associated_person test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "associated_person01", "associated_person02", "associated_person03", "employee01", "employee02", "employee03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("KOTA_TEST_ASSOCIATED_PERSON_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["KOTA_TEST_ASSOCIATED_PERSON_ENTID"] = idmap,
    ["KOTA_TEST_LIVE"] = "FALSE",
    ["KOTA_TEST_EXPLAIN"] = "FALSE",
    ["KOTA_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["KOTA_TEST_ASSOCIATED_PERSON_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["employee_id"] == nil then
    idmap_resolved["employee_id"] = idmap_resolved["employee01"]
  end

  if env["KOTA_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["KOTA_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["KOTA_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["KOTA_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/associated_person_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/kota-sdk/go"
	"github.com/voxgig-sdk/kota-sdk/go/core"

	vs "github.com/voxgig-sdk/kota-sdk/go/utility/struct"
)

func TestAssociatedPersonEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.AssociatedPerson(nil)
		if ent == nil {
			t.Fatal("expected non-nil AssociatedPersonEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"associated_person": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.AssociatedPerson(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.AssociatedPerson(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := associated_personBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "associated_person." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set KOTA_TEST_ASSOCIATED_PERSON_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		associatedPersonRef01Ent := client.AssociatedPerson(nil)
		associatedPersonRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "associated_person"}, setup.data), "associated_person_ref01"))
		associatedPersonRef01Data["employee_id"] = setup.idmap["employee01"]

		associatedPersonRef01DataResult, err := associatedPersonRef01Ent.Create(associatedPersonRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		associatedPersonRef01Data = core.ToMapAny(entityData(associatedPersonRef01DataResult))
		if associatedPersonRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if associatedPersonRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		associatedPersonRef01Match := map[string]any{
			"employee_id": setup.idmap["employee01"],
		}

		associatedPersonRef01ListResult, err := associatedPersonRef01Ent.List(associatedPersonRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		associatedPersonRef01List, associatedPersonRef01ListOk := associatedPersonRef01ListResult.([]any)
		if !associatedPersonRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", associatedPersonRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(associatedPersonRef01List), map[string]any{"id": associatedPersonRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		associatedPersonRef01DataUp0Up := map[string]any{
			"id": associatedPersonRef01Data["id"],
			"employee_id": setup.idmap["employee_id"],
		}

		associatedPersonRef01MarkdefUp0Name := "date_of_birth"
		associatedPersonRef01MarkdefUp0Value := fmt.Sprintf("Mark01-associated_person_ref01_%d", setup.now)
		associatedPersonRef01DataUp0Up[associatedPersonRef01MarkdefUp0Name] = associatedPersonRef01MarkdefUp0Value

		associatedPersonRef01ResdataUp0Result, err := associatedPersonRef01Ent.Update(associatedPersonRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		associatedPersonRef01ResdataUp0 := core.ToMapAny(entityData(associatedPersonRef01ResdataUp0Result))
		if associatedPersonRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if associatedPersonRef01ResdataUp0["id"] != associatedPersonRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if associatedPersonRef01ResdataUp0[associatedPersonRef01MarkdefUp0Name] != associatedPersonRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", associatedPersonRef01MarkdefUp0Name, associatedPersonRef01ResdataUp0[associatedPersonRef01MarkdefUp0Name])
		}

		// LOAD
		associatedPersonRef01MatchDt0 := map[string]any{
			"id": associatedPersonRef01Data["id"],
		}
		associatedPersonRef01DataDt0Loaded, err := associatedPersonRef01Ent.Load(associatedPersonRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		associatedPersonRef01DataDt0LoadResult := core.ToMapAny(entityData(associatedPersonRef01DataDt0Loaded))
		if associatedPersonRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if associatedPersonRef01DataDt0LoadResult["id"] != associatedPersonRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		associatedPersonRef01MatchRm0 := map[string]any{
			"id": associatedPersonRef01Data["id"],
		}
		_, err = associatedPersonRef01Ent.Remove(associatedPersonRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		associatedPersonRef01MatchRt0 := map[string]any{
			"employee_id": setup.idmap["employee01"],
		}

		associatedPersonRef01ListRt0Result, err := associatedPersonRef01Ent.List(associatedPersonRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		associatedPersonRef01ListRt0, associatedPersonRef01ListRt0Ok := associatedPersonRef01ListRt0Result.([]any)
		if !associatedPersonRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", associatedPersonRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(associatedPersonRef01ListRt0), map[string]any{"id": associatedPersonRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func associated_personBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "associated_person", "AssociatedPersonTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read associated_person test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse associated_person test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"associated_person01", "associated_person02", "associated_person03", "employee01", "employee02", "employee03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("KOTA_TEST_ASSOCIATED_PERSON_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"KOTA_TEST_ASSOCIATED_PERSON_ENTID": idmap,
		"KOTA_TEST_LIVE":      "FALSE",
		"KOTA_TEST_EXPLAIN":   "FALSE",
		"KOTA_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["KOTA_TEST_ASSOCIATED_PERSON_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add employee_id alias for update test.
	if idmapResolved["employee_id"] == nil {
		idmapResolved["employee_id"] = idmapResolved["employee01"]
	}

	if env["KOTA_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["KOTA_APIKEY"],
			},
			extra,
		})
		client = sdk.NewKotaSDK(core.ToMapAny(mergedOpts))
	}

	live := env["KOTA_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["KOTA_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
